### DISCUSSION TITLE
Error Running Tests After Cloning Code

### DISCUSSION THREAD

**Discussion by stefkuk** [triplea-game/triplea#12423] [New Contributor]
I'm trying to add a small feature and I want to test my implementation but I'm running into errors. Even without changes to the code, when I download the latest repo and follow the Developer Setup instructions and run HeadedGameRunner in IntelliJ, I get the error:

Task :game-app:map-data:compileJava FAILED. class com.sun.tools.javac.tree.JCTree$JCLiteral cannot be cast to class com.sun.tools.javac.tree.JCTree$JCCaseLabel (com.sun.tools.javac.tree.JCTree$JCLiteral and com.sun.tools.javac.tree.JCTree$JCCaseLabel are in module jdk.compiler of loader 'app'.

Any idea how to get HeadedGameRunner to run out of the box?

**Comment by DanVanAtta** [comment]
- Can you build via CLI?
- Which OS are you using?
- How exactly are you running `HeadedGameRunner.java` from intellij? 
   - can you confirm that intellij is using gradle or whether it is compiling it?  

That is in a setting:
![image](https://github.com/triplea-game/triplea/assets/12397753/794829b4-11fe-40ea-bcbb-b6b25f0c0aca)


- Does intellij see the project as compiled? 
- Is there more to the stack-trace?
- Is your Gradle JVM set to Java-11? (configured on the same screenshot above)
- In 'File > project structure" do you have Java-11 configured?
![image](https://github.com/triplea-game/triplea/assets/12397753/238cfe72-7c1e-4323-b26a-50032922da88)

**Comment by stefkuk** [comment]
Hey Dan, thanks for getting back to me! I've seen your name everywhere in the code.

That was super thorough. My issue was the Gradle JVM version that defaulted to 20 instead of 11. Downloading gradle 11 and running that got it working.

Even though I now have it running, I'm still not sure how to test my changes. I wanted to fix the error that happens when you try to open a saved game via mac OS. Current behavior is loading the save game and throwing an error message. Instead, I want it to start the game. How can I open a tsvg on my computer and have it open in the Intellij HeadedGameRunner?

I can submit a PR so you can see my suggested changes

**Reply by DanVanAtta** [reply]
@RoiEXLab  or @asvitkine  might have additional suggestions for how to test that update.
My understanding is there is a file association that gets registered with the OS, at that point the OS knows to invoke 'TripleA' when a .tsvg file is opened. Getting the OS to invoke something that is in turn run by IntelliJ seems infeasible.

I would try to split this problem up in two as much as possible. First, does the OS invoke TripleA, and if so - what do the program arguments look like? Second, given a set of args to TripleA - does TripleA do the right thing?

For the OS registration part:
(A)  It is possible to create a shadowJar (an executable Jar file), and perhaps you can configure the OS to invoke that Jar file. 
(B)  It is possible to do a local build where you create an install4j bundle and could then install. 

B* is slow, and I recently try to do a full release build from CLI & kept crashing due to out-of-heap memory. There's a config somewhere that needs to be adjusted which I have not found. This might not be a clean path, in addition to the install4j bundling takes several minutes to finish. We might also need to set you up with the install4j license key as well if we go down this path.

(C) Commit something that would give you diagnostic info when the application is invoked. Essentially add some logging to the application start, commit & merge that & then use the installers built as part of the mainline (master) build. This approach would have the slowest iteration speed.

> I can submit a PR so you can see my suggested changes

Draft PRs early & often is generally a good thing. The more opportunity for feedback, usually the better. It is a bit of a developer anti-pattern to "retreat into a cave" and then re-emerge a few months later with some giant merge request. Though, my availability is somewhat shaky, my responses can be delayed and that in itself can slow things down. Which all the more makes a draft-PR a good thing - all the more people that can provide feedback.

--------------

Worth double checking, I don't know for sure, but there might have been some significant progress in this area since 2.5; I'm not sure what the status of this problem is relative to 2.6.  Have you installed a 2.6 version and verified it still have the problem?

**Reply by RoiEXLab** [reply]
> My understanding is there is a file association that gets registered with the OS, at that point the OS knows to invoke 'TripleA' when a .tsvg file is opened. Getting the OS to invoke something that is in turn run by IntelliJ seems infeasible.

Yes and no. On Windows and Linux when double clicking on a file it will just start a new process with the file to be opened as CLI parameter based on a registry entry (windows) or a special config file (most common Linux distributions, google freedesktop.org if you're interested in details). On MacOS however this works differently. Instead of starting a new process every time, it starts a process just normally if no other process has been started yet, and simply invokes a special handler on a process that is already running. This is where the `Info.plist` and [`Desktop#openFileHandler`](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/Desktop.html#setOpenFileHandler(java.awt.desktop.OpenFilesHandler)) come into play. So the CLI arguments are completely irrelevant for this feature on MacOS as far as I'm aware of.

Anyways to cite the [javadocs](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/Desktop.html#setOpenFileHandler(java.awt.desktop.OpenFilesHandler)):
> Please note that for macOS, notifications are only sent if the Java app is a bundled application, with a CFBundleDocumentTypes array present in its Info.plist. Check the[ Apple Developer Documentation](https://developer.apple.com/documentation) for more information about Info.plist.

Going down this rabbit hole I found this documentation: https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html
Which seems to describe some of the relevant keys that need to be present to trigger the file open handler at all.

So my approach to test stuff would look as follows:
- In the main function create a new thread that sleeps for a couple of seconds and then runs the listener being passed to [`Desktop#openFileHandler`](https://docs.oracle.com/en/java/javase/21/docs/api/java.desktop/java/awt/Desktop.html#setOpenFileHandler(java.awt.desktop.OpenFilesHandler)).
- If this works as expected and opens a file normally then the problem is likely related to the `Info.plist` configuration, so the handler is invoked in an unexpected way.
- In case it doesn't I don't know enough how to create a proper `Info.plist` file. Not sure what a minimal "bundled application" looks like. But if you can install TripleA under the assumption that the current code works as it should, you should be able to just modify TripleA's `Info.plist` file to get it working.

Of course once you tested everything like this on it's own there's probably no way around installing Install4j and building a package on your own. Hope this helps a little bit at least

**Comment by asvitkine** [comment]
@skuklinsky 

Given the error message you were looking for doesn't exist in 2.6, are you sure the issue you're trying to fix isn't already fixed in 2.6?
Have you tested with a 2.6 prerelease?

You can get the latest one here:
https://github.com/triplea-game/triplea/releases

But even if that issue is fixed, we'd certainly be happy to have more contributors. There are some other issues with save game files that could use some help! Let us know!

**Comment by stefkuk** [comment]
In 2.6 prerelease this scenario no longer throws an error, so I will put this issue to bed.

I was originally thinking that clicking a tsvg from the OS should also immediately launch into the local game, but now I think that opening to the home page panel (current behavior) is ideal since there are multiple options other than Start Local Game (like host networked game).

Thanks for all the input.

I'll let you know if I try to tackle the other save game issues.

**Reply by DanVanAtta** [reply]
Sounds good!

**Reply by asvitkine** [reply]
Here are the other saved game issues that still exist that need help, if you're interested:
https://github.com/triplea-game/triplea/issues/10523
https://github.com/triplea-game/triplea/issues/11665

**Reply by asvitkine** [reply]
And another: https://github.com/triplea-game/triplea/issues/11257
